### ISSUE TITLE
Main thread is blocked on concurrent async operations

### ISSUE THREAD

**Issue by fantix** [sqlalchemy/sqlalchemy#5581]
**Describe the bug**

SQLAlchemy < 1.4 is sometimes using `threading.Lock()` to control concurrency, and this is also used in 1.4 AsyncEngine / AsyncConnection. As the 2nd `threading.Lock.acquire()` in a row from the same thread will block the main thread, this leads to freezing bugs in e.g. concurrent calls of event `first_connect()`, please see more details below.

**Expected behavior**

Async concurrency should work without blocking the main thread. Typically this involves using `asyncio.Lock()` to control concurrency.

**To Reproduce**

```py
import asyncio
from sqlalchemy.ext.asyncio import create_async_engine


async def main():
    e = create_async_engine("postgresql+asyncpg:///")
    c1 = e.connect()
    c2 = e.connect()
    await asyncio.wait([c1, c2])


asyncio.run(main())
```

**Error**

It blocks forever but issuing keyboard interrupt generates:

```
^CTraceback (most recent call last):
  File "concurrent_async_connect.py", line 12, in <module>
    asyncio.run(main())
  File "python3.8/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "python3.8/asyncio/base_events.py", line 603, in run_until_complete
    self.run_forever()
  File "python3.8/asyncio/base_events.py", line 570, in run_forever
    self._run_once()
  File "python3.8/asyncio/base_events.py", line 1859, in _run_once
    handle._run()
  File "python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "python3.8/asyncio/tasks.py", line 684, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "sqlalchemy/lib/sqlalchemy/ext/asyncio/engine.py", line 79, in start
    self.sync_connection = await (greenlet_spawn(self.sync_engine.connect))
  File "sqlalchemy/lib/sqlalchemy/util/_concurrency_py3k.py", line 93, in greenlet_spawn
    result = context.switch(value)
  File "sqlalchemy/lib/sqlalchemy/future/engine.py", line 426, in connect
    return super(Engine, self).connect()
  File "sqlalchemy/lib/sqlalchemy/engine/base.py", line 2914, in connect
    return self._connection_cls(self, close_with_result=close_with_result)
  File "sqlalchemy/lib/sqlalchemy/engine/base.py", line 90, in __init__
    else engine.raw_connection()
  File "sqlalchemy/lib/sqlalchemy/engine/base.py", line 2993, in raw_connection
    return self._wrap_pool_connect(self.pool.connect, _connection)
  File "sqlalchemy/lib/sqlalchemy/engine/base.py", line 2960, in _wrap_pool_connect
    return fn()
  File "sqlalchemy/lib/sqlalchemy/pool/base.py", line 301, in connect
    return _ConnectionFairy._checkout(self)
  File "sqlalchemy/lib/sqlalchemy/pool/base.py", line 698, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "sqlalchemy/lib/sqlalchemy/pool/base.py", line 419, in checkout
    rec = pool._do_get()
  File "sqlalchemy/lib/sqlalchemy/pool/impl.py", line 142, in _do_get
    self._dec_overflow()
  File "sqlalchemy/lib/sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.raise_(
  File "sqlalchemy/lib/sqlalchemy/util/compat.py", line 179, in raise_
    raise exception
  File "sqlalchemy/lib/sqlalchemy/pool/impl.py", line 139, in _do_get
    return self._create_connection()
  File "sqlalchemy/lib/sqlalchemy/pool/base.py", line 247, in _create_connection
    return _ConnectionRecord(self)
  File "sqlalchemy/lib/sqlalchemy/pool/base.py", line 362, in __init__
    self.__connect(first_connect_check=True)
  File "sqlalchemy/lib/sqlalchemy/pool/base.py", line 588, in __connect
    pool.dispatch.first_connect.for_modify(
  File "sqlalchemy/lib/sqlalchemy/event/attr.py", line 316, in exec_once_unless_exception
    self._exec_once_impl(True, *args, **kw)
  File "sqlalchemy/lib/sqlalchemy/event/attr.py", line 284, in _exec_once_impl
    with self._exec_once_mutex:
KeyboardInterrupt
```

**Versions.**
 - OS: Darwin Kernel Version 19.5.0
 - Python: 3.8.2
 - SQLAlchemy: 1.4.0b1 @ a2f2863f5af934a94ed96539e852cb874c3203c0
 - Database: PostgreSQL 12.4
 - DBAPI: asyncpg 0.21.0

**Additional context**

https://github.com/sqlalchemy/sqlalchemy/blob/a2f2863f5af934a94ed96539e852cb874c3203c0/lib/sqlalchemy/event/attr.py#L277-L294

At line 284, the 1st coroutine owned the lock and proceeded on to execute async queries in `first_connect()` (context-switched to the event loop), and the 2nd coroutine came in and blocked the whole main thread trying to acquire the same lock twice.

As a proof-of-concept, using `asyncio.Lock()` solves the problem:

```python
class _CompoundListener(_InstanceLevelDispatch):
    __slots__ = "_exec_once_mutex", "_exec_once"

    def _memoized_attr__exec_once_mutex(self):
        import asyncio
        return asyncio.Lock()

    def _exec_once_impl(self, retry_on_exception, *args, **kw):
        from sqlalchemy.util import await_only

        await_only(self._exec_once_mutex.acquire())
        try:
            if not self._exec_once:
                try:
                    self(*args, **kw)
                    exception = False
                except:
                    exception = True
                    raise
                finally:
                    if not exception or not retry_on_exception:
                        self._exec_once = True
        finally:
            self._exec_once_mutex.release()
```

But obviously this is not a proper fix, and I don't have much idea for a proper fix atm. For this particular issue, some async-adaption changes may be needed for the event system. I haven't tried to find problematic scenarios for other `threading.Lock()` usages, but theoretically they are at risk too.

**Have a nice day!**

**Comment by zzzeek** [comment]
so great you are early testing this.   yes there's a mutex stuck in there and there is also one in mapper.py.    The mutex stuck here, we can make that parameterized since it's part of the event listener that is local to the pool which knows its an async pool.     the one in mapper.py I'm not so sure about, it can be avoided by calling configure_mappers() before the program runs.

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **master** branch:

**Adapt event exec_once_mutex to asyncio** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2231

**Comment by zzzeek** [comment]
apparently an asyncio.Lock() binds itself to the current event loop when you create it so I might need to do this more carefully

**Comment by zzzeek** [comment]
So it looks like with asyncio objects like Queue, Lock, etc., these objects all assume they are created in a particular thread, and then within that thread there will be exactly one event loop, and that's the loop they bind towards forever.   so we can assume Python asyncio is not compatible with any system that wants to do something like, have multiple event loops running in individual threads, right?   i think twisted tried to do that?     basically if you say create_async_engine() in a thread, the connection pool gets created, it has an asyncio.Queue inside of it, that queue is welded to the event loop of the current thread.   your SQLAlchemy engine can never be used inside any other thread with any other kind of event loop.   is that all good assumptions to make?

**Comment by zzzeek** [comment]
this basically implies that I can't use any module-level "asyncio" locks.   Since the lock used for mapper.configure_mappers() doesn't actually run any IO within the mapper configure process, it seems this would be safe.  a user can of course try to do IO inside of a mapper configure event hook, that would be extremely strange, so for now if they were to try to do that while using asyncio, they'd be out of luck.

**Comment by fantix** [comment]
Thank you for the fast fix! I can confirm that it's working, and it's a good fix.

> So it looks like with asyncio objects like Queue, Lock, etc., these objects all assume they are created in a particular thread, and then within that thread there will be exactly one event loop, and that's the loop they bind towards forever.

Yes, this is mostly the behavior of such asyncio objects when created without a currently-running event loop. Since Python 3.5.3, people could dynamically create - let's say - locks in a thread that has more than one event loop. [`get_event_loop()`](https://github.com/python/cpython/blob/51ba5b7d0c194b0c2b1e5d647e70e3538b8dde3e/Lib/asyncio/events.py#L659-L671) will automatically pick up the currently running loop, but once created the lock is bound to that loop forever.

> so we can assume Python asyncio is not compatible with any system that wants to do something like, have multiple event loops running in individual threads, right?

Well not necessarily - as far as users are consistent with the loops. For example, I can do this without issues now:

```python
import asyncio

from sqlalchemy.ext.asyncio import create_async_engine


async def get_engine():
    return create_async_engine("postgresql+asyncpg:///")


async def concurrent_first_connect(e):
    c1 = e.connect()
    c2 = e.connect()
    await asyncio.wait([c1, c2])


loop1 = asyncio.new_event_loop()
loop2 = asyncio.new_event_loop()
e1 = loop1.run_until_complete(get_engine())
e2 = loop2.run_until_complete(get_engine())
loop1.run_until_complete(concurrent_first_connect(e1))
loop2.run_until_complete(concurrent_first_connect(e2))
```

And I believe even if the loops are in different threads, it should also work. But if the user messes up with the loops, it will fail. For example, using `e2` in `loop1` won't work:

```python
loop1.run_until_complete(concurrent_first_connect(e2))
```

Similarly, this won't work too:

```python
loop1 = asyncio.new_event_loop()
loop2 = asyncio.new_event_loop()
e1 = create_async_engine("postgresql+asyncpg:///")
e2 = create_async_engine("postgresql+asyncpg:///")
loop1.run_until_complete(concurrent_first_connect(e1))
loop2.run_until_complete(concurrent_first_connect(e2))
```

Because both `e1`/`e2` are bound to the default event loop which is neither `loop1` nor `loop2`.

However, these are the users' fault, not the library's.

> i think twisted tried to do that?

I don't have a clue because my outdated Twisted knowledge is over 10 years old.

> basically if you say create_async_engine() in a thread, the connection pool gets created, it has an asyncio.Queue inside of it, that queue is welded to the event loop of the current thread. your SQLAlchemy engine can never be used inside any other thread with any other kind of event loop. is that all good assumptions to make?

Yes, this is a valid assumption if `create_async_engine()` is called in a thread without a running event loop.

I'd re-phrase the assumption as, "if you say `create_async_engine()` in an event loop, every asyncio object used by this SQLAlchemy engine is attached to that particular event loop, and this engine can never be used with any other kind of event loop". In addition, if there is no event loop in a thread, `create_async_engine()` will attach to the **only implicitly-created default** event loop of that thread.

> this basically implies that I can't use any module-level "asyncio" locks.

You're right, module-level "asyncio" locks won't work.

> Since the lock used for mapper.configure_mappers() doesn't actually run any IO within the mapper configure process, it seems this is safe to keep as a regular threading.Lock(). a user can of course try to do IO inside of a mapper configure event hook, that would be extremely strange, so for now if they were to try to do that while using asyncio, they'd be out of luck.

Sounds fair to me. Btw as GINO depends on SQLAlchemy Core only, I'm not actually testing the async SQLAlchemy ORM very much.

**Comment by JamzumSum** [comment]
Sorry I'm not familiar with the inner logic of sqlalchemy, but why is the return object in `_memoized_attr__exec_once_mutex` still `threading.Lock`?

https://github.com/sqlalchemy/sqlalchemy/blob/77fc8216a74e6b2d0efc6591c6c735687bd10002/lib/sqlalchemy/event/attr.py#L418-L431

I've observed getattr to `_ListenerCollection(...)._exec_once_mutex` leads to a func call to `_CompoundListener(...)._memoized_attr__exec_once_mutex`, which returns a `threading.Lock` and then deadlock my asyncio app.

I modify the return val of `_memoized_attr__exec_once_mutex` like @fantix does in https://github.com/sqlalchemy/sqlalchemy/issues/5581#issue-700744521
and it works on my tests. Is this modify correct?
``` py
class _CompoundListener(_InstanceLevelDispatch):
    __slots__ = "_exec_once_mutex", "_exec_once", "_exec_w_sync_once"

    def _set_asyncio(self):
        self._exec_once_mutex = AsyncAdaptedLock()

    def _memoized_attr__exec_once_mutex(self):
        return AsyncAdaptedLock()
```

**Comment by CaselIT** [comment]
No, there should no need to do that. Please open a new issue with a proper example.

**Comment by zzzeek** [comment]
> Sorry I'm not familiar with the inner logic of sqlalchemy, but why is the return object in `_memoized_attr__exec_once_mutex` still `threading.Lock`?

`_memoized_attr__exec_once_mutex` is only called if the object does not have an "_exec_once_mutex" attribute set on it.  once something sets `_exec_once_mutex`, then `_memoized_attr__exec_once_mutex` is never called.  

The Lock returned by the `_memoized_attr__exec_once_mutex` method is the default style of lock returned if asyncio is not in use.  if asyncio is in use, the [set_asyncio](https://github.com/sqlalchemy/sqlalchemy/blob/d7caa3c0bef64c710e45ed646756da65c4241bdd/lib/sqlalchemy/event/attr.py#L273) method sets `_exec_once_mutex` to be an AsyncAdaptedLock, and `_memoized_attr__exec_once_mutex` will never be called.

> I've observed getattr to `_ListenerCollection(...)._exec_once_mutex` leads to a func call to `_CompoundListener(...)._memoized_attr__exec_once_mutex`, which returns a `threading.Lock` and then deadlock my asyncio app.

this would indicate your application is not establishing the correct asyncio dependencies in place.  if you use create_async_engine() with the correct DBAPI driver, there should be no issue.   I'm not familiar with how your issue might be reproduced so I can't provide any more detail.

> 
> I modify the return val of `_memoized_attr__exec_once_mutex` like @fantix does in [#5581 (comment)](https://github.com/sqlalchemy/sqlalchemy/issues/5581#issue-700744521) and it works on my tests. Is this modify correct?

no, that's incorrect.
